third_party/move/move-bytecode-verifier/src/signature_v2.rs

```diff
@@ -110,148 +110,142 @@ impl<const N: usize> BitsetTypeParameterConstraints<N> {
-                required_abilities.requires(),
-                param_constraints,
-            )?;
-        },
-        Function(_, _, abilities) => {
-            assert_abilities(*abilities, required_abilities)?;
-            // Note we do not need to check abilities of argument or result types,
-            // they do not matter for the `required_abilities`.
-        },
-        Struct(sh_idx) => {
-            let handle = &struct_handles[sh_idx.0 as usize];
-            assert_abilities(handle.abilities, required_abilities)?;
-        },
-        StructInstantiation(sh_idx, ty_args) => {
-            let handle = &struct_handles[sh_idx.0 as usize];
-
-            assert_abilities(handle.abilities, required_abilities)?;
+                ),
+                        ),
+                    )
+                } else {
+                    Ok(())
+                }
+            };
 
-            // TODO: is this needed?
-            if handle.type_parameters.len() != ty_args.len() {
-                return Err(
-                    PartialVMError::new(StatusCode::NUMBER_OF_TYPE_ARGUMENTS_MISMATCH)
-                        .with_message(format!(
-                            "expected {} type argument(s), got {}",
-                            handle.type_parameters.len(),
-                            ty_args.len()
-                        )),
-                );
-            }
-            for (ty_param, ty_arg) in handle.type_parameters.iter().zip(ty_args.iter()) {
-                let required_abilities = if ty_param.is_phantom {
-                    ty_param.constraints
+        match ty {
+            TypeParameter(param_idx) => param_constraints.insert(*param_idx, required_abilities),
+            U8 | U16 | U32 | U64 | U128 | U256 | Bool | Address => {
+                assert_abilities(AbilitySet::PRIMITIVES, required_abilities)?;
+            },
+            Signer => {
+                assert_abilities(AbilitySet::SIGNER, required_abilities)?;
+            },
+            Reference(ty) | MutableReference(ty) => {
+                if allow_ref {
+                    assert_abilities(AbilitySet::REFERENCES, required_abilities)?;
+                    self.check_ty(ty, false, AbilitySet::EMPTY, param_constraints)?;
                 } else {
-                    ty_param.constraints.union(required_abilities.requires())
-                };
+                    return Err(PartialVMError::new(StatusCode::INVALID_SIGNATURE_TOKEN)
+                        .with_message("reference not allowed".to_string()));
+                }
+            },
+            Vector(ty) => {
+                assert_abilities(AbilitySet::VECTOR, required_abilities)?;
+                self.check_ty(ty, false, required_abilities.requires(), param_constraints)?;
+            },
+            Function(params, results, abilities) => {
+                assert_abilities(*abilities, required_abilities)?;
+                if self.sig_checker_v2_fix_function_signatures {
```

Comment:
Note: only changed this part, the rest of changes on Github are due to moving functions into `impl` block